kotlinx-coroutines-core/jvm/src/CoroutineContext.kt

```diff
@@ -10,35 +10,84 @@ import kotlin.coroutines.jvm.internal.CoroutineStackFrame
+    return foldCopies(this, addedContext, false)
+}
+
+private val hasCopyableElements: (Boolean, CoroutineContext.Element) -> Boolean = { result, it ->
+    result || it is CopyableThreadContextElement<*>
+}
+
+/**
+ * Folds two contexts properly applying [CopyableThreadContextElement] rules when necessary.
+ * The rules are the following:
+ * * If both context do not have CTCE, the sum of two contexts is returned
+ * * Every CTCE from left-hand side context that does not have matching (by key) element from right-hand side context
+ *   is [copied][CopyableThreadContextElement.copyForChild]
+ * * Every CTCE from left-hand side context that has matching element in right-hand side context is [merged][CopyableThreadContextElement.mergeForChild]
+ * * Every CTCE from right-hand side context that hasn't been merged is copied
+ * * Everything else is added to the resulting context as is.
+ */
+private fun foldCopies(originalContext: CoroutineContext, appendContext: CoroutineContext, isNewCoroutine: Boolean): CoroutineContext {
+    // Do we have something to copy left-hand side?
+    val hasElementsLeft = originalContext.fold(false, hasCopyableElements)
+    val hasElementsRight = appendContext.fold(false, hasCopyableElements)
+
+    // Nothing to fold, so just return the sum of contexts
+    if (!hasElementsLeft && !hasElementsRight) {
+        return originalContext + appendContext
     }
-    if (!hasToCopy) return this
-    return fold<CoroutineContext>(EmptyCoroutineContext) { combined, it ->
-        combined + if (it is CopyableThreadContextElement<*>) it.copyForChildCoroutine() else it
+
+    var leftoverContext = appendContext
+    val folded = originalContext.fold<CoroutineContext>(EmptyCoroutineContext) { result, element ->
+        if (element !is CopyableThreadContextElement<*>) return@fold result + element
+        // Will this element be overwritten?
+        val newElement = leftoverContext[element.key]
+        // No, just copy it
+        if (newElement == null) {
+            // For 'withContext'-like builders we do not copy as the element is not shared
+            return@fold result + if (isNewCoroutine) element.copyForChild() else element
+        }
+        // Yes, then first remove the element from append context
+        leftoverContext = leftoverContext.minusKey(element.key)
+        // Return the sum
+        @Suppress("UNCHECKED_CAST")
+        return@fold result + (element as CopyableThreadContextElement<Any?>).mergeForChild(newElement)
+    }
+
+    if (hasElementsRight) {
+        leftoverContext = leftoverContext.fold<CoroutineContext>(EmptyCoroutineContext) { result, element ->
+            // We're appending new context element -- we have to copy it, otherwise it may be shared with others
+            if (element is CopyableThreadContextElement<*>) {
+                return@fold result + element.copyForChild()
+            }
+            return@fold result + element
```

Comment:
I believe the original version is more readable

Reply:
You're right, for some reason I can't place a breakpoint on block subexpressions here. How about this then?
```suggestion
            result + if (element is CopyableThreadContextElement<*>) {
                element.copyForChild()
            } else {
                element
            }
```
If not, feel free to just disregard this.

Reply:
In general, I prefer
```
if (condition) {
    return foo
}
doOrReturnOtherFoo
```

over
```
return if (condition) {
    foo
} else {
   otherFoo
}
```

to keep the number of nested blocks minimal. This is, of course, opinionated, so we can stick to the most "concise" one or to the proposed middle-ground if you do object

Reply:
In general, I'm the opposite and will typically prefer the other way because it highlights the symmetry between the cases and simplifies mental parsing of the control flow. In particular, in this case, I certainly don't find nesting to be an issue.

In any case, we're splitting hairs here. Both versions are perfectly fine, I think.